clang/include/clang/Sema/Sema.h

```diff
@@ -1759,6 +1759,34 @@ class Sema final {
     }
   };
 
+  /// Bitmask to contain the list of reasons a single diagnostic should be
+  /// emitted, based on its language.  This permits multiple offload systems
+  /// to coexist in the same translation unit.
+  enum class DeviceDiagnosticReason : unsigned char {
+    /// Diagnostic doesn't apply to anything. Included for completeness, but
+    /// should make this a no-op.
+    None = 0,
+    /// OpenMP specific diagnostic.
+    OmpDevice = 1 << 0,
+    OmpHost = 1 << 1,
+    OmpAll = OmpDevice | OmpHost,
+    /// CUDA specific diagnostics.
+    CudaDevice = 1 << 2,
+    CudaHost = 1 << 3,
+    CudaAll = CudaDevice | CudaHost,
+    /// SYCL specific diagnostic.
+    Sycl = 1 << 4,
+    /// ESIMD specific diagnostic.
+    Esimd = 1 << 5,
+    /// A flag representing 'all'.  This can be used to avoid the check
+    /// all-together and make this behave as it did before the
+    /// DiagnosticReason was added (that is, unconditionally emit).
+    /// Note: This needs to be updated if any flags above are added.
+    All = 0x3F,
+
+    LLVM_MARK_AS_BITMASK_ENUM(/*LargestValue=*/All)
```

Comment:
Is 'All' the right answer here?  Based on the documentation, it implies to me it should be Esimd.  @AaronBallman : Do you know more about this that you can clarify?

Reply:
Researching, it seems that AMDGPU/SIMemoryLeagalizer.cpp uses this with an 'all' that is OK.  

Though it reminds me, should 'All' be defined as the 'or-ing' of all the above instead of an integral value?  @AaronBallman this is a style issue.

Reply:
I waffled on whether it should be an `|` or not and can go either way, but I think it's probably better to define it as `OmpAll | CudaAll | Sycl | Esimd` to be very clear on the intent.

Reply:
Changed